template_reports/office_renderer/loops.py

```diff
@@ -0,0 +1,235 @@
+"""
+Functions to handle loop functionality in PPTX templates.
+"""
+
+import re
+from typing import Dict, List, Optional, Tuple, Any
+from copy import deepcopy
+
+from ..templating import resolve_tag
+
+# Patterns for loop directives
+LOOP_START_PATTERN = re.compile(r"%\s*loop\s+(\w+)\s+in\s+(.+?)\s*%")
+LOOP_END_PATTERN = re.compile(r"%\s*endloop\s*%")
+
+
+def extract_loop_directive(text: str | None) -> tuple[str | None, str | None]:
+    """Return (variable, collection) if *text* contains a loop directive."""
+    if not text:
+        return None, None
+    
+    match = LOOP_START_PATTERN.search(text.strip())
+    if match:
+        variable = match.group(1)
+        collection = match.group(2)
+        return variable, collection
+    
+    return None, None
+
+
+def is_loop_start(shape) -> bool:
+    """Return True if the shape text indicates a loop start."""
+    if not hasattr(shape, "text_frame") or not hasattr(shape.text_frame, "text"):
+        return False
+    
+    variable, collection = extract_loop_directive(shape.text_frame.text)
+    return variable is not None and collection is not None
+
+
+def is_loop_end(shape) -> bool:
+    """Return True if the shape text indicates a loop end."""
+    if not hasattr(shape, "text_frame") or not hasattr(shape.text_frame, "text"):
+        return False
+    
+    text = shape.text_frame.text
+    if not text:
+        return False
+    
+    return LOOP_END_PATTERN.search(text.strip()) is not None
+
+
+def deep_copy_slide(presentation, slide_index, slide_dict=None):
```

Comment:
delete deep_copy_slide. i have instructions later for how and where to handle duplication.